### DISCUSSION TITLE
Proposal: Extend `cover` entity to support tilt only devices

### DISCUSSION THREAD

**Discussion by raman325** [home-assistant/architecture#908] [Entity Models]
## Background

The Z-Wave and Matter specifications include cover like devices that only support tilt (they do not support an overall position). Per an internal discussion, the best way today to represent this device type is to use a number entity.

## Options on how to handle

1. Use `number` entity platform
2. Extend `cover` entity platform
3. Create new entity platform

## Order of preference

2, 1, 3

## Advantages of using a `number` entity over a `cover` entity

- fits this use case using an existing entity type - no new entity platform code needs to be added
- relatively easy to implement setting tilt position using the number set_value service

## Advantages of using a `cover` entity over a `number` entity

- Tilting covers can be open or closed, and `number` entities have no way to represent that
- The cover entity supports device classes that are more in line with what a user expects from a device that supports cover tilt
- Just by example, the use case fits my mental model for what a cover is. Example: I have venetian blinds with a motor that controls the tilt. The blinds themselves are a cover device, it just happens to only support tilt.
- `number` entities have no way to stop a position change whereas `cover` supports that

CC @marcelveldt @frenck 

## Proposal

- Add a `tilt_state` attribute (`tilt_is_closed`? open to ideas here).
- State falls back to `tilt_state` when `is_closed` is `None` and `tilt` is a supported feature
- ~~Add an additional feature for `CLOSE_TILT_ALTERNATE` which allows for a secondary tilt (blind slats can close in two directions)~~
- ~~Consider adding an optional `direction` attribute with an enum that's either `UP_AND_DOWN` or `LEFT_AND_RIGHT`. This can be used by the frontend to decide how to display the `close_tilt` and `close_tilt_alternate` service~~
- ~~add a `close_tilt_alternate` service or update the `close_tilt` service to accept an optional parameter that specifies direction somehow~~

EDIT: per discussion, this proposal will focus on tilt only and not the bidrectional piece

**Comment by Jc2k** [comment]
Can you sketch out what changes you thing are needed for 2? Can't you just ignore "tilt" and just use the main open/closed/position states? The fact that they phsyically "tilt" is just an implementation detail of your cover isn't it? And the main state is how open the blind is? Or am I missing something?

Would (1) just be the % open or closed?

**Reply by raman325** [reply]
> Can you sketch out what changes you thing are needed for 2? Can't you just ignore "tilt" and just use the main open/closed/position states? The fact that they phsyically "tilt" is just an implementation detail of your cover isn't it? And the main state is how open the blind is? Or am I missing something?

That's a fair question/comment. If that's an acceptable solution then this proposal can be closed. My understanding is that covers were never intended to have only tilt functionality, hence the discussion. I don't have an alternate suggestion on how to extend covers to support tilt only devices but your suggestion seems reasonable.

> Would (1) just be the % open or closed?

I guess that's part of the problem, you can use any min and max for a number entity. In the case of Z-Wave, it would be 0-99 where 0 = 0% and 99 = 100%. Not very intuitive in my opinion

**Comment by parautenbach** [comment]
Another issue is that some blinds can tilt in two directions, and the direction you tilt it in matters. In other words, the middle position would be 0% tilt, and thus you could have -100% or 100% tilt. Of course, one would not like to cause a breaking change in this regard, so supporting functionality such as a tilt direction could be added.

**Reply by raman325** [reply]
good point. I did not mention this but this is indeed an issue even with using the tilt functionality in cover. Right now I have hardcoded the close to be in a single direction but it would be ideal if a user could choose the direction

**Comment by starkillerOG** [comment]
I find CLOSE_TILT_ALTERNATE a bit vage, I would suggest CLOSE_TILT_BIDIRECTIONAL or CLOSE_TILT_BOTH_DIRRCTIONS or something.

**Comment by starkillerOG** [comment]
Could we also consider honeycomb covers also referred to as Top Down Bottom Up covers?
They have 2 bars with fabric in between. If the top bar goes to the top and the bottom bar to the bottom the window is covers, if both bars come together it is open.

![Screenshot_20230527_074938_Samsung Internet](https://github.com/home-assistant/architecture/assets/14811189/5757da73-955c-4be2-a03c-1db9da394f90)

Currently I implemented them as 3 cover entities, 1 for the top, one for the bottom and one to move both. See the documentation here:
https://www.home-assistant.io/integrations/motion_blinds/#top-down-bottom-up-tdbu-blinds

**Reply by raman325** [reply]
FWIW for z-wave covers that also support this the plan was to do the same thing as you, create 3 cover entities

**Comment by vexofp** [comment]
I am interested in working to improve the tilt-only cover situation a bit, but I think some additional discussion might be necessary as to the best direction to go.

# Problems

First, a quick overview of the big things that are broken/bad with the current state of affairs:
1. Tilt-only covers always report a state of "Unknown"
2. They do not work with most scripts/groups/voice assistants/etc. that leverage the main open/close methods/services

# Solutions

After some brainstorming, I see a few different potential ways to go about improving the situation:

Note that the chosen solution here may also have impacts for related proposals such as #912.

## Option 1: Report `OPEN`/`CLOSE`/`STOP`/`SET_POSITION` features and implement associated methods/services

The entity would report that it supports `OPEN`/`CLOSE`/`SET_POSITION` and implement the associated methods in terms of tilting/rotating instead of raising/lowering. This handily fixes both of the problems above.

This would require tweaks to any integrations providing such covers today. It has a few sub-options we would need to decide between:

### Option 1A: ... and do _not_ report `OPEN_`/`CLOSE_`/`STOP_`/`SET_TILT_POSITION` nor implement the tilt methods/services

This is is what @Jc2k [suggested above](https://github.com/home-assistant/architecture/discussions/908#discussioncomment-5994080), I think.

Such entities would now no longer support the tilt-specific methods/services, which feels a bit strange given that they do in fact tilt. This would also be a breaking change for existing devices.

### Option 1B: ... and _also_ report `OPEN_`/`CLOSE_`/`STOP_`/`SET_TILT_POSITION` and implement the tilt methods/services

This would make the two "sets" of methods/services do the same thing. E.g. `open_cover()` and `open_cover_tilt()` would be the same.

This is backwards compatible, but likely to confuse anything that thinks the two are independent and should be independently controllable. For example, the front-end would display two control sliders (one for position, one for tilt), which actually both control the same thing. This leads to a bad user experience.

One option might be to add some sort of `POSITION_IS_TILT` feature for this case. This would let front-ends know that they only need to show one control.

### Option 1C: ... and do _not_ report `OPEN_`/`CLOSE_`/`STOP_`/`SET_TILT_POSITION` but _do_ still implement the tilt methods/services

It feels odd to implement methods/services, which by the reported features should never be called. But this would potentially make the change less breaking, while otherwise being mostly like 1A.

## Option 2: Report only `*_TILT` features, and have `state` fall back to reporting tilt state

This is closer to what @raman325 [suggested above](https://github.com/home-assistant/architecture/discussions/908#discussion-5230037).

For entities that do not support positioning, their entity state is defined by their tilt state instead. This solves only problem 1, however.

This solution maintains the most accurate representation of the entity, in terms of which features it does and does not physically support.

### Option 2A: ... and nothing else

Problem 2 above is not fixed yet:
* Assistants can be individually fixed to support mapping their semantic open/close to the tilt method/services instead.
* Would we be okay with mapping "close" on a group of covers to "close tilt", for the tilt-only covers?
* How would scripts handle this case?

### Option 2B: ... and proxy `open_cover()` to `open_cover_tilt()`, etc.

This would provide an implementation of the "positional" methods/services for tilt-only covers, despite the fact that the entity does not actually advertise support for them. This is like option 1C, but in reverse.

This _partially_ fixes problem 2 above. Assistants or anything that is checking the features should never actually call these methods/services, so they still would still need manual work to add explicit support for mapping open/close/position to tilt.

# Preference

To me, 1A (or 1C for more backwards compatibility) seems simplest. This does not actually involve any architecture change; just (breaking) changes to any such integrations.

2B also seems okay. A bit more complex, but only requires changes in `cover`, instead of potentially many integrations. It also avoids completely losing the information that the entity tilts, instead of moves.

1B is viable, but definitely the most work and riskiest. Especially since front-ends would need to immediately understand the new feature or else display bad controls.
